[Ingest pipelines] Add ability to stop pipeline simulation

alisonelizabeth opened the pull request:
This PR adds the ability to stop/reset testing a pipeline.

This addresses feedback received here: https://github.com/elastic/kibana/pull/74964#pullrequestreview-470502194.

### How to test
1. Create a pipeline and add some sample documents (for a sample pipeline and documents, see: https://github.com/elastic/kibana/pull/74964).
2. Click the documents dropdown, and click the "Reset" link.
3. Verify modal opens
4. After confirming, verify the sample documents added in step 1 no longer exist, and the status of each processor is reset to "Not run".

### Screenshots

<img width="457" alt="Screen Shot 2020-09-22 at 1 20 13 PM" src="https://user-images.githubusercontent.com/5226211/93915719-6bccb480-fcd6-11ea-83ad-2bf545b90c69.png">

<img width="558" alt="Screen Shot 2020-09-22 at 1 20 19 PM" src="https://user-images.githubusercontent.com/5226211/93915722-6c654b00-fcd6-11ea-8d70-31bed0605463.png">

elasticmachine commented:
Pinging @elastic/es-ui (Team:Elasticsearch UI)

kibanamachine commented:
## :green_heart: Build Succeeded
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/76261/)
* Commit: 743c91594a00668785ec478f66eb9a3bf315029c

### Build metrics
#### @kbn/optimizer bundle module count
| id | [value](https://github.com/elastic/kibana/commit/743c91594a00668785ec478f66eb9a3bf315029c) | diff | [baseline](https://github.com/elastic/kibana/commit/6d819b7a1d95e8ecc50eef364ef1448544bf21da) |
| --- | --- | --- | --- |
| ingestPipelines | 448 | +1 | 447 |

#### async chunks size
| id | [value](https://github.com/elastic/kibana/commit/743c91594a00668785ec478f66eb9a3bf315029c) | diff | [baseline](https://github.com/elastic/kibana/commit/6d819b7a1d95e8ecc50eef364ef1448544bf21da) |
| --- | --- | --- | --- |
| ingestPipelines | 769.0KB | +3.4KB | 765.6KB |

To update your PR or re-run it, just comment with:
`@elasticmachine merge upstream`

<!--PIPELINE
{"builds":[{"status":"SUCCESS","url":"https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/76261/","number":"76261","commit":"743c91594a00668785ec478f66eb9a3bf315029c"}]}
PIPELINE-->

jloleysens commented on x-pack/plugins/ingest_pipelines/public/application/components/pipeline_processors_editor/components/test_pipeline/test_pipeline_actions.tsx:
```diff
@@ -49,6 +49,12 @@ export const TestPipelineActions: FunctionComponent = () => {
   };
 
+  const stopPipelineSimulation = () => {
+    setCurrentTestPipelineData({
```
nit; I think a better name for `setCurrentTestPipelineData` would be `testPipelineDataDispatch` (if we want to disambiguate) otherwise `dispatch`.

jloleysens reviewed (approved):
Great work @alisonelizabeth ! I think this is a nice UX polish! I looked at the code and everything looks good to me, thanks for adding tests too!

I looked at the current UI and I have some suggestions about what I think would be cool, but not necessarily work that needs to be done in this PR.

To my mind the "edit documents" and "reset" buttons represent actions that both are scoped to documents, so it makes sense they live together as they currently do. An improvement on this, I think, is moving them out of the document selector dropdown -- so that the selector is just a selector (second image).

The controls for selecting and managing documents can be grouped visually (per image 1, not showing view output as I think this is distinct from managing documents). These controls could live in a bottom bar because they are not "primary" level actions to editing the pipeline but relevant to the process of testing.

*Image 1*
<img width="1186" alt="Screenshot 2020-09-23 at 12 44 15" src="https://user-images.githubusercontent.com/8155004/94003618-c67c1400-fd9b-11ea-8c9e-aecdbc22289f.png">

*Image 2*
<img width="940" alt="Screenshot 2020-09-23 at 12 42 25" src="https://user-images.githubusercontent.com/8155004/94003749-f4f9ef00-fd9b-11ea-9291-3dc64ca8732a.png">

I think it would be great to hear from @mdefazio or others on this too.

mdefazio commented:
One thought I had on this was instead of 'Reset', this is a toggle button that says 'Autorun test' and then 'Test paused'.  (Probably better text here).  Then reseting the documents can be done in the document flyout only.

alisonelizabeth commented:
> One thought I had on this was instead of 'Reset', this is a toggle button that says 'Autorun test' and then 'Test paused'. (Probably better text here). Then reseting the documents can be done in the document flyout only.

@mdefazio Thanks for the suggestions! Would you mind taking another look when you get a chance?

Based on your feedback on https://github.com/elastic/kibana/pull/77939, I've added a "Clear all" button to the flyout and removed the "Reset" button originally added in this PR. The "paused" toggle is an interesting idea. I think it needs to be fleshed out a little more. For example, if it was "paused", would the documents in the dropdown list be disabled? Also, would you still be able to click the "View output" button?

I think adding the "Clear all" button resolves the original concern about being stuck in testing mode and is a good step forward. I think I'd prefer to hold off on the toggle for now, unless you feel strongly about it. 

Latest screenshots:

<img width="480" alt="Screen Shot 2020-09-28 at 11 53 18 AM" src="https://user-images.githubusercontent.com/5226211/94455621-4b42a580-0181-11eb-8667-5c375bf175d8.png">
<img width="557" alt="Screen Shot 2020-09-28 at 11 53 29 AM" src="https://user-images.githubusercontent.com/5226211/94455623-4b42a580-0181-11eb-9583-a99e6ff5a256.png">
<img width="525" alt="Screen Shot 2020-09-28 at 11 53 33 AM" src="https://user-images.githubusercontent.com/5226211/94455624-4bdb3c00-0181-11eb-8252-4de61b3fa1a1.png">

kibanamachine commented:
## :green_heart: Build Succeeded
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/77474/)
* Commit: 39c7d795a21bc1460091fe422b08ca735d85527e

### Metrics [[docs](https://www.elastic.co/guide/en/kibana/master/ci-metrics.html)]
#### @kbn/optimizer bundle module count
| id | [value](https://github.com/elastic/kibana/commit/39c7d795a21bc1460091fe422b08ca735d85527e) | diff | [baseline](https://github.com/elastic/kibana/commit/88df93bed56ea02c9a84f78eac32c9653829c5eb) |
| --- | --- | --- | --- |
| ingestPipelines | 493 | +10 | 483 |

#### async chunks size
| id | [value](https://github.com/elastic/kibana/commit/39c7d795a21bc1460091fe422b08ca735d85527e) | diff | [baseline](https://github.com/elastic/kibana/commit/88df93bed56ea02c9a84f78eac32c9653829c5eb) |
| --- | --- | --- | --- |
| ingestPipelines | 815.1KB | +6.4KB | 808.7KB |

### History
* :green_heart: [Build #76261](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/76261/) succeeded 743c91594a00668785ec478f66eb9a3bf315029c

To update your PR or re-run it, just comment with:
`@elasticmachine merge upstream`

<!--PIPELINE
{"builds":[{"status":"SUCCESS","url":"https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/76261/","number":"76261","commit":"743c91594a00668785ec478f66eb9a3bf315029c"},{"status":"SUCCESS","url":"https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/77474/","number":"77474","commit":"39c7d795a21bc1460091fe422b08ca735d85527e"}]}
PIPELINE-->

kibanamachine commented:
## :green_heart: Build Succeeded
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/77518/)
* Commit: f5205b85fbb2e3f95719add4a2428d29ee194334

### Metrics [[docs](https://www.elastic.co/guide/en/kibana/master/ci-metrics.html)]
#### @kbn/optimizer bundle module count
| id | [value](https://github.com/elastic/kibana/commit/f5205b85fbb2e3f95719add4a2428d29ee194334) | diff | [baseline](https://github.com/elastic/kibana/commit/88df93bed56ea02c9a84f78eac32c9653829c5eb) |
| --- | --- | --- | --- |
| ingestPipelines | 493 | +10 | 483 |

#### async chunks size
| id | [value](https://github.com/elastic/kibana/commit/f5205b85fbb2e3f95719add4a2428d29ee194334) | diff | [baseline](https://github.com/elastic/kibana/commit/88df93bed56ea02c9a84f78eac32c9653829c5eb) |
| --- | --- | --- | --- |
| ingestPipelines | 814.9KB | +6.2KB | 808.7KB |

### History
* :green_heart: [Build #77474](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/77474/) succeeded 39c7d795a21bc1460091fe422b08ca735d85527e
* :green_heart: [Build #76261](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/76261/) succeeded 743c91594a00668785ec478f66eb9a3bf315029c

To update your PR or re-run it, just comment with:
`@elasticmachine merge upstream`

<!--PIPELINE
{"builds":[{"status":"SUCCESS","url":"https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/76261/","number":"76261","commit":"743c91594a00668785ec478f66eb9a3bf315029c"},{"status":"SUCCESS","url":"https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/77474/","number":"77474","commit":"39c7d795a21bc1460091fe422b08ca735d85527e"},{"status":"SUCCESS","url":"https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/77518/","number":"77518","commit":"f5205b85fbb2e3f95719add4a2428d29ee194334"}]}
PIPELINE-->

mdefazio reviewed (changes requested):
This all looks good.  I have one tiny nit: Can we update the Edit documents button so it uses the `EuiPopoverFooter` (and add the `fullWidth` prop?  Might simplify things a bit but also want to make sure the button spans the width of the popover.

The toggle isn't really necessary anymore IMO.

alisonelizabeth commented:
Thanks @mdefazio! I was not aware of the `EuiPopoverFooter`. That does simplify things a bit! Mind taking one last look when you get a chance?

<img width="387" alt="Screen Shot 2020-09-28 at 9 20 04 PM" src="https://user-images.githubusercontent.com/5226211/94501814-b23b7b00-01d0-11eb-8dd0-d37342bdfae4.png">

kibanamachine commented:
## :green_heart: Build Succeeded
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/77715/)
* Commit: 2b5bb3bbf13cb0209294492a53fb1f093cab2791

### Metrics [[docs](https://www.elastic.co/guide/en/kibana/master/ci-metrics.html)]
#### @kbn/optimizer bundle module count
| id | [value](https://github.com/elastic/kibana/commit/2b5bb3bbf13cb0209294492a53fb1f093cab2791) | diff | [baseline](https://github.com/elastic/kibana/commit/88df93bed56ea02c9a84f78eac32c9653829c5eb) |
| --- | --- | --- | --- |
| ingestPipelines | 493 | +10 | 483 |

#### async chunks size
| id | [value](https://github.com/elastic/kibana/commit/2b5bb3bbf13cb0209294492a53fb1f093cab2791) | diff | [baseline](https://github.com/elastic/kibana/commit/88df93bed56ea02c9a84f78eac32c9653829c5eb) |
| --- | --- | --- | --- |
| ingestPipelines | 814.4KB | +5.7KB | 808.7KB |

### History
* :green_heart: [Build #77518](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/77518/) succeeded f5205b85fbb2e3f95719add4a2428d29ee194334
* :green_heart: [Build #77474](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/77474/) succeeded 39c7d795a21bc1460091fe422b08ca735d85527e
* :green_heart: [Build #76261](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/76261/) succeeded 743c91594a00668785ec478f66eb9a3bf315029c

To update your PR or re-run it, just comment with:
`@elasticmachine merge upstream`

<!--PIPELINE
{"builds":[{"status":"SUCCESS","url":"https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/76261/","number":"76261","commit":"743c91594a00668785ec478f66eb9a3bf315029c"},{"status":"SUCCESS","url":"https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/77474/","number":"77474","commit":"39c7d795a21bc1460091fe422b08ca735d85527e"},{"status":"SUCCESS","url":"https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/77518/","number":"77518","commit":"f5205b85fbb2e3f95719add4a2428d29ee194334"},{"status":"SUCCESS","url":"https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/77715/","number":"77715","commit":"2b5bb3bbf13cb0209294492a53fb1f093cab2791"}]}
PIPELINE-->

mdefazio reviewed (approved):
Looks good! Thanks for making that edit.
